server/routes/experiments/biomed_nl/traversal.py

```diff
@@ -0,0 +1,536 @@
+    '''Filters paths based on cosine similarity between property descriptions 
+    and the query.
+
+    Args:
+        pct: The percentage of top properties to keep (e.g., 0.1 for 10%).
+
+    Returns:
+        A list of the top properties that were selected for filtering.
+    '''
+
+    property_descriptions = self.path_store.get_property_descriptions()
+
+    # TODO: add a description to description in DC KG.
+    property_descriptions['description'] = (
+        'The description describes the entity by its characteristics or '
+        'attributes, providing information about that entity to help to define '
+        'and distinguish it.')
+
+    prop_embeds = {
+        prop:
+            self.embeddings_model.encode(property_descriptions[prop],
+                                         convert_to_tensor=True)
+        for prop in property_descriptions
+    }
+    query_embed = self.embeddings_model.encode(self.query,
+                                               convert_to_tensor=True)
+
+    properties_to_similarity_score = {
+        prop: util.cos_sim(query_embed, prop_embed).item()
+        for prop, prop_embed in prop_embeds.items()
+    }
+    num_top_properties = math.ceil(len(properties_to_similarity_score) * pct)
+    top_props = [
+        property[0]
+        for property in sorted(properties_to_similarity_score.items(),
+                               key=lambda item: item[1],
+                               reverse=True)[:num_top_properties]
+    ]
+
+    self.path_store.filter_by_properties(top_props)
+
+    return top_props
+
+  def select_paths(self):
+    '''Selects paths to purpose using the Gemini model based on a prompt.
+
+    Returns:
+        True if the traversal should terminate, False otherwise.
+    '''
+
+    should_terminate = True
+
+    prompt = utils.TRAVERSAL_PROMPT.format(
+        QUERY=self.query,
+        START_ENT=self.start_entity_name,
+        START_DCIDS=self.start_dcids,
+        LINKS=self.path_store.get_paths_from_start(),
+        METADATA=self.path_store.get_property_descriptions())
+    response = self.gemini.models.generate_content(model=self.gemini_model_str,
+                                                   contents=prompt)
+    input_tokens, output_tokens = utils.get_gemini_response_token_counts(
+        response)
+    self.input_tokens += input_tokens
+    self.output_tokens += output_tokens
+
+    if response.text.startswith('DONE'):
+      should_terminate = True
+    elif response.text.startswith('CONTINUE'):
+      should_terminate = False
+    elif response.text.startswith('NONE'):
+      return True
+    else:
+      # This branch only occurs if there has been an error with Gemini.
+      return True
+
+    paths = [
+        line.strip()
+        for line in response.text.split('\n')
+        if bool(re.match(r"^path\d+:", line.strip()))
+    ]
+
+    self.selected_paths = PathStore.from_selected_paths(paths, self.path_store)
```

Comment:
curious why we need to save this instead of just returning it?

Reply:
updated this so selected_paths are saved in the path_store. do you think that's an improvement?